src/main/java/conversion/BaseServlet.java

```diff
@@ -189,6 +190,26 @@ private String getFileName(final Part part) {
         return null;
     }
 
+    /**
+     * Gets the full URL before the part containing the path(s) specified in
+     * urlPatterns of the servlet.
+     *
+     * @param request
+     * @return protocol://servername/contextPath
+     */
+    protected static String getContextURL(final HttpServletRequest request) {
+        StringBuffer full = request.getRequestURL();
+        String context = request.getContextPath();
+        // If context does not exist.
+        if (context.isEmpty()) {
+            context = request.getHeader("Host");
+        }
+        // Get url up to and including context.
+        int index = full.indexOf(context);
+        index += context.length();
+        return full.substring(0, index);
```

Comment:
request.getHeader("Host") may cause an issue as it would not include the protocol.

Would this work?

```java
final StringBuffer full = request.getRequestURL();
return full.substring(0, full.length() - request.getServletPath().length());
```

Reply:
When I've been testing, request.getHeader("Host") has been returning the protocol (seemed strange to me too that it would do that). Though I've tested your solution and it's much more concise so we can go with that.